### ISSUE TITLE
Commit ec720ea breaks mitogen @ ansible 2.5.0

### ISSUE THREAD

**Issue by mlaurense** [dw/mitogen#189]
I just updated mitogen by a git pull, but immediately it stopped working. The culprit seems to be commit ec720eab1a1d29102fde2ecdc6addaa69bc33dd8; reverting to the commit before it is all working fine.

```
TASK [Gathering Facts] **********************************************************************************************************************************************************************************************************************************************
An exception occurred during task execution. To see the full traceback, use -vvv. The error was:     .format(row, col, self.prev_row, self.prev_col))
fatal: [production999.local]: FAILED! => {"msg": "Unexpected failure during module execution.", "stdout": ""}
```

Verbose output:
```
ansible-playbook 2.5.0
  config file = ansible.cfg
  configured module search path = [u'/Users/mike/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /Library/Python/2.7/site-packages/ansible
  executable location = /usr/local/bin/ansible-playbook
  python version = 2.7.10 (default, Jul 15 2017, 17:16:57) [GCC 4.2.1 Compatible Apple LLVM 9.0.0 (clang-900.0.31)]
Using ansible.cfg as config file
Parsed production999.local:2222, inventory source with host_list plugin
statically imported: roles/geerlingguy.nginx/tasks/setup-RedHat.yml
statically imported: roles/geerlingguy.nginx/tasks/setup-Ubuntu.yml
statically imported: roles/geerlingguy.nginx/tasks/setup-Debian.yml
statically imported: roles/geerlingguy.nginx/tasks/setup-FreeBSD.yml
statically imported: roles/geerlingguy.nginx/tasks/setup-OpenBSD.yml
statically imported: roles/geerlingguy.nginx/tasks/setup-Archlinux.yml
statically imported: roles/geerlingguy.nginx/tasks/vhosts.yml

PLAYBOOK: base.yml ******************************************************************************************************************************************************************************************************************************************
1 plays in playbooks/base.yml

PLAY [all] **********************************************************************************************************************************************************************************************************************************************************
[pid 16972] 16:56:16.366943 D ansible_mitogen.process: Service pool configured: size=16

TASK [Gathering Facts] **********************************************************************************************************************************************************************************************************************************************
task path: playbooks/base.yml:1
[pid 16973] 16:56:16.405159 D ansible_mitogen.mixins: _remote_expand_user(u'~/.ansible/tmp', sudoable=True)
[pid 16973] 16:56:16.406791 D mitogen: unix.connect(path='/var/folders/yw/bnfr5phn3c50hdpqtw36ddym0000gn/T/mitogen_unix_2mJI9k')
[pid 16973] 16:56:16.408343 D mitogen: unix.connect(): local ID is 1, remote is 0
[pid 16972] 16:56:16.411818 D mitogen: mitogen.ssh.Stream('default').connect()
[pid 16973] 16:56:16.442464 D mitogen: mitogen.core.Stream('unix_listener.16972').on_disconnect()
[pid 16972] 16:56:16.443023 D mitogen: mitogen.core.Stream('unix_client.16973').on_disconnect()
[pid 16973] 16:56:16.443238 D mitogen: Waker(Broker(0x109701950) rfd=44, wfd=45).on_disconnect()
The full traceback is:
Traceback (most recent call last):
  File "/Library/Python/2.7/site-packages/ansible/executor/task_executor.py", line 138, in run
    res = self._execute()
  File "/Library/Python/2.7/site-packages/ansible/executor/task_executor.py", line 554, in _execute
    result = self._handler.run(task_vars=variables)
  File "mitogen-master/ansible_mitogen/mixins.py", line 114, in run
    return super(ActionModuleMixin, self).run(tmp, task_vars)
  File "/Library/Python/2.7/site-packages/ansible/plugins/action/normal.py", line 46, in run
    result = merge_hash(result, self._execute_module(task_vars=task_vars, wrap_async=wrap_async))
  File "mitogen-master/ansible_mitogen/mixins.py", line 322, in _execute_module
    remote_tmp=mitogen.utils.cast(self._get_remote_tmp()),
  File "mitogen-master/ansible_mitogen/mixins.py", line 190, in _get_remote_tmp
    return self._remote_expand_user(s)
  File "mitogen-master/ansible_mitogen/mixins.py", line 292, in _remote_expand_user
    return os.path.join(self._connection.homedir, path[2:])
  File "mitogen-master/ansible_mitogen/connection.py", line 133, in homedir
    self._connect()
  File "mitogen-master/ansible_mitogen/connection.py", line 277, in _connect
    self.host, self._homedir = self._connect_ssh()
  File "mitogen-master/ansible_mitogen/connection.py", line 206, in _connect_ssh
    for term in shlex.split(s or '')
  File "mitogen-master/ansible_mitogen/connection.py", line 164, in _wrap_connect
    kwargs=mitogen.utils.cast(kwargs),
  File "mitogen-master/mitogen/service.py", line 344, in call
    return recv.get().unpickle()
  File "mitogen-master/mitogen/core.py", line 360, in unpickle
    raise obj
CallError: exceptions.ValueError: start (784,0) precedes previous end (784,1)
  File "mitogen-master/mitogen/service.py", line 251, in _on_receive_message
    response = getattr(self, method_name)(**kwargs)
  File "mitogen-master/ansible_mitogen/services.py", line 102, in connect
    context = method(**kwargs)
  File "mitogen-master/mitogen/parent.py", line 870, in ssh
    return self.connect('ssh', **kwargs)
  File "mitogen-master/mitogen/parent.py", line 840, in connect
    return self._connect(klass, name=name, **kwargs)
  File "mitogen-master/mitogen/parent.py", line 826, in _connect
    stream.connect()
  File "mitogen-master/mitogen/ssh.py", line 114, in connect
    super(Stream, self).connect()
  File "mitogen-master/mitogen/parent.py", line 584, in connect
    self.pid, fd = self.start_child()
  File "mitogen-master/mitogen/parent.py", line 580, in start_child
    return self.create_child(*self.get_boot_command())
  File "mitogen-master/mitogen/ssh.py", line 110, in get_boot_command
    base = super(Stream, self).get_boot_command()
  File "mitogen-master/mitogen/parent.py", line 537, in get_boot_command
    preamble_compressed = self.get_preamble()
  File "mitogen-master/mitogen/parent.py", line 574, in get_preamble
    return zlib.compress(minimize_source(source), 9)
  File "mitogen-master/mitogen/compat/functools.py", line 236, in wrapper
    result = user_function(*args, **kwds)
  File "mitogen-master/mitogen/parent.py", line 100, in minimize_source
    return tokenize.untokenize(tokens)
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/tokenize.py", line 268, in untokenize
    return ut.untokenize(iterable)
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/tokenize.py", line 208, in untokenize
    self.add_whitespace(start)
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/tokenize.py", line 190, in add_whitespace
    .format(row, col, self.prev_row, self.prev_col))


fatal: [production999.local]: FAILED! => {
    "msg": "Unexpected failure during module execution.", 
    "stdout": ""
}
        to retry, use: --limit @playbooks/base.retry

PLAY RECAP **********************************************************************************************************************************************************************************************************************************************************
production999.local        : ok=0    changed=0    unreachable=0    failed=1   


real    0m1.130s
user    0m0.809s
sys     0m0.277s
[pid 16972] 16:56:16.514106 E mitogen: mitogen.service.Pool(0x1096e6d10, size=16, th='mitogen.service.Pool.1096e6d10.worker-2'): channel or latch closed, exitting: Channel closed by local end.
[pid 16972] 16:56:16.514760 E mitogen: mitogen.service.Pool(0x1096e6d10, size=16, th='mitogen.service.Pool.1096e6d10.worker-1'): channel or latch closed, exitting: Channel closed by local end.
[pid 16972] 16:56:16.515131 E mitogen: mitogen.service.Pool(0x1096e6d10, size=16, th='mitogen.service.Pool.1096e6d10.worker-3'): channel or latch closed, exitting: Channel closed by local end.
[pid 16972] 16:56:16.516015 E mitogen: mitogen.service.Pool(0x1096e6d10, size=16, th='mitogen.service.Pool.1096e6d10.worker-4'): channel or latch closed, exitting: None
[pid 16972] 16:56:16.516452 E mitogen: mitogen.service.Pool(0x1096e6d10, size=16, th='mitogen.service.Pool.1096e6d10.worker-5'): channel or latch closed, exitting: None
[pid 16972] 16:56:16.516875 E mitogen: mitogen.service.Pool(0x1096e6d10, size=16, th='mitogen.service.Pool.1096e6d10.worker-6'): channel or latch closed, exitting: None
[pid 16972] 16:56:16.517287 E mitogen: mitogen.service.Pool(0x1096e6d10, size=16, th='mitogen.service.Pool.1096e6d10.worker-7'): channel or latch closed, exitting: None
[pid 16972] 16:56:16.517662 E mitogen: mitogen.service.Pool(0x1096e6d10, size=16, th='mitogen.service.Pool.1096e6d10.worker-8'): channel or latch closed, exitting: None
[pid 16972] 16:56:16.518310 E mitogen: mitogen.service.Pool(0x1096e6d10, size=16, th='mitogen.service.Pool.1096e6d10.worker-10'): channel or latch closed, exitting: None
[pid 16972] 16:56:16.518767 E mitogen: mitogen.service.Pool(0x1096e6d10, size=16, th='mitogen.service.Pool.1096e6d10.worker-12'): channel or latch closed, exitting: None
[pid 16972] 16:56:16.519224 E mitogen: mitogen.service.Pool(0x1096e6d10, size=16, th='mitogen.service.Pool.1096e6d10.worker-9'): channel or latch closed, exitting: None
[pid 16972] 16:56:16.519637 E mitogen: mitogen.service.Pool(0x1096e6d10, size=16, th='mitogen.service.Pool.1096e6d10.worker-11'): channel or latch closed, exitting: None
[pid 16972] 16:56:16.520072 E mitogen: mitogen.service.Pool(0x1096e6d10, size=16, th='mitogen.service.Pool.1096e6d10.worker-13'): channel or latch closed, exitting: None
[pid 16972] 16:56:16.520491 E mitogen: mitogen.service.Pool(0x1096e6d10, size=16, th='mitogen.service.Pool.1096e6d10.worker-14'): channel or latch closed, exitting: None
[pid 16972] 16:56:16.520984 E mitogen: mitogen.service.Pool(0x1096e6d10, size=16, th='mitogen.service.Pool.1096e6d10.worker-15'): channel or latch closed, exitting: None
[pid 16972] 16:56:16.521400 E mitogen: mitogen.service.Pool(0x1096e6d10, size=16, th='mitogen.service.Pool.1096e6d10.worker-0'): channel or latch closed, exitting: None
[pid 16972] 16:56:16.521975 D mitogen: Waker(Broker(0x1096e65d0) rfd=41, wfd=43).on_disconnect()
[pid 16972] 16:56:16.522431 D mitogen: <mitogen.unix.Listener object at 0x1096e6a50>.on_disconnect()
```

**Comment by dw** [comment]
Sorry for the trouble, and thanks for catching this :)  What
OS/distribution is the controller machine?

**Comment by dw** [comment]
It's a Python bug that was fixed in 2.7.11: https://bugs.python.org/issue20387

Thanks for reporting

**Comment by dw** [comment]
Gloriously, it's just a 16 line function we need to override in a subclass if Python<2.7.11. I'll have this fixed shortly. Thanks again.
